README.md
<a href="https://travis-ci.org/GreyRook/Draw2Script"><img src="https://travis-ci.org/GreyRook/Draw2Script.svg?branch=master" align=right></a>

# Draw2Script
Export Source Code from Adobe Illustrator.  Inspired by [DrawScript](http://drawscri.pt/)

## Features

 * Supported Export Formats
   * EaselJS tiny API
   * PIXI_tiny API
   * PIXI_graphics
   * JSON
   
## Screenshot

![Ellipse PIXI_tiny](docs/draw2script_screenshot_ellipse_pixi_tiny.png)

## Installation using Extension Manager (Adobe CC2014 and CS6)
 * Install [Extension Manager](https://www.adobe.com/exchange/em_download/)
 * Run Draw2Script.zxp
 * If you have both CS6 and CC2014 installed together, Draw2Script may use the wrong version of Adobe Extension Manager and install the plugin for CS6 instead of CC2014, or vice versa
 
## Installation (Adobe CC2015)

 * Download [Zip](https://github.com/GreyRook/Draw2Script/archive/master.zip)
 * Extract Zip to
   * System extension folder
     * Win: C:\Program Files (x86)\Common Files\Adobe\CEP\extensions
     * Mac: /Library/Application Support/Adobe/CEP/extensions   
   * or Per-user extension folder
     * Win: C:\Users\\[username]\AppData\Roaming\Adobe\CEP\extensions
	 * Mac: ~/Library/Application Support/Adobe/CEP/extensions
  
## State

Beta State.


CHANGELOG.md
# ROADMAP
 * CompoundPath Support

# History
* 2015-07-13: Alpha release 0.0.1
* 2015-07-15: Alpha release 0.0.2
  * Support for [PIXI_tiny](https://github.com/GreyRook/PIXI_tiny)
  * RGB Color Support
* 2015-07-17: Alpha release 0.0.3
  * Support for JSON
  * Better UI
  * Error handling for "nothing selected" and "overflow"
* 2015-07-20: Beta release 0.1.07
  * Support for PIXI_graphics
  * Unit tests



test/draw2script.js
testGetColor();
testConvertBitToBase64();
testConvertToHex();
testConvertNumberToBits();

function testGetColor() {
	var testColor = {
		typename: 'RGBColor',
		red: '0',
		green: '0',
		blue: '0'
	};
	
	var testOpactiy = 100;
	expect(getColor(testColor, testOpactiy)).to.eql("'rgba(0,0,0,1)'");	
	
	testColor = {
			typename: 'RGBColor',
			red: '255',
			green: '255',
			blue: '255'
	};
	expect(getColor(testColor, testOpactiy)).to.eql("'rgba(255,255,255,1)'");	
	
	testColor = {
			typename: 'CMYKColor',
			cyan: '0',
			black: '100',
			magenta: '0',
			yellow: '0'
	}
	expect(getColor(testColor, testOpactiy)).to.eql("'rgba(0,0,0,1)'");

	hexColor = true;
	expect(getColor(testColor, testOpactiy)).to.eql("0x000000");
}

function testConvertBitToBase64() {
	var bits = ["000000", "000001", "111111000000"];
	var base64 = ["A", "B", "/A"];
	for(var i = 0; i < bits.length; i++) {
		expect(convertBitToBase64(bits[i])).to.eql(base64[i]);
	}
}

function testConvertToHex() {
	var number = [0,2,5,6,255,100];
	var hex = ["00","02","05","06","ff","64"];
	for(var i = 0; i < number.length; i++) {
		expect(convertToHex(number[i])).to.eql(hex[i]);
	}
}

function testConvertNumberToBits() {
	var number = [0,0.2,0.4,0.6,0.8,10.0,100.0,200000.0,-0.1,-2.0];
	var bits12 = ["000000000000","000000000010","000000000100","000000000110","000000001000",
	              "000001100100","001111101000","overflow","100000000001","100000010100"];
	var bits18 = ["000000000000000000","000000000000000010","000000000000000100",
	              "000000000000000110","000000000000001000","000000000001100100",
	              "000000001111101000","overflow","100000000000000001","100000000000010100"];
	for(var i = 0; i < number.length; i++) {
		expect(convertNumberToBits(number[i],12)).to.eql(bits12[i]);
		expect(convertNumberToBits(number[i],18)).to.eql(bits18[i]);
	}
}

spec/createjs.spec.js
// hack for executing code on node.js
if(typeof module !== 'undefined') {
    fs = require('fs');
    var code = fs.readFileSync('js/parser/createjs.js', 'utf-8');
    eval(code);
}


// actual test-case
describe("Createjs Parser Test Suite", function() {

	var createjs = new Createjs();
	
	it("testGenerate", function() {
		var pathItem = {
	        "typename" : "PathItem",
	        "closed" : true,
	        "filled" : true,
	        "fillColor" : {
	            "red" : 0,
				"green" : 0,
				"blue" : 0,
	            "typename" : "RGBColor"
	        },
	        "opacity" : 100,
	        "stroked" : false,
	        "strokeColor" : {
	            "cyan" : 0,
	            "magenta" : 0,
	            "yellow" : 0,
	            "black" : 100,
	            "typename" : "CMYKColor"
	        },
	        "strokeDashes" : [

	        ],
	        "strokeDashOffset" : 0,
	        "strokeCap" : {},
	        "strokeJoin" : {},
	        "strokeMiterLimit" : 10,
	        "strokeWidth" : 6,
	        "pathPoints" : [{
	                "anchor" : [
	                    0,
	                    0
	                ],
	                "leftDirection" : [
	                    0,
	                    0
	                ],
	                "rightDirection" : [
	                    0,
	                    0
	                ]
	            }, {
	                "anchor" : [
	                    10,
	                    0
	                ],
	                "leftDirection" : [
	                    10,
	                    0
	                ],
	                "rightDirection" : [
	                    10,
	                    0
	                ]
	            }, {
	                "anchor" : [
	                    10,
	                    10
	                ],
	                "leftDirection" : [
	                    10,
	                    10
	                ],
	                "rightDirection" : [
	                    10,
	                    10
	                ]
	            }, {
	                "anchor" : [
	                    0,
						10
	                ],
	                "leftDirection" : [
	                     0,
						10
	                ],
	                "rightDirection" : [
	                     0,
						10
	                ]
	            }
	        ]
	    };
		
		var groupItem = {
			"typename": "GroupItem",
			"groupItems": [],
			"pathItems": [pathItem],
			"compoundPathItems": []
		};
		
		var json = {
			"cropBox" : [0,0],
			"selection" : [groupItem]
		}
		
		createjs.cropBox = json.cropBox;
		
		var expectedResult = ".f('rgba(0,0,0,1)').p('" + 
			createjs.convertPathItemToPathInstruction(pathItem) + "').ef().cp()"
		this.cropBox = json.cropBox;
		var result = createjs.generate(json);
		expect(result).toBe(expectedResult);

	});

	it("testConvertPathItemToPathInstruction", function() {
		createjs.cropBox = [0,0];
		var pathItem = {
	        "pathPoints" : [{
	                "anchor" : [
	                    0,
	                    0
	                ],
	                "leftDirection" : [
	                    0,
	                    0
	                ],
	                "rightDirection" : [
	                    0,
	                    0
	                ]
	            }
	        ]
	    };
		expect(createjs.convertPathItemToPathInstruction(pathItem)).toBe("EAAAAAA");
		
		pathItem = {
		    "pathPoints" : [{
		            "anchor" : [
		                0,
		                0
		            ],
		            "leftDirection" : [
		                0,
		                0
		            ],
		            "rightDirection" : [
		                0,
		                0
		            ]
		        }, {
		            "anchor" : [
		                10,
		                0
		            ],
		            "leftDirection" : [
		                10,
		                0
		            ],
		            "rightDirection" : [
		                10,
		                0
		            ]
		        }
		    ]
		};
		expect(createjs.convertPathItemToPathInstruction(pathItem)).toBe("EAAAAAAMABkAAA");
	});
	
    it("testGetColor", function () {
	    var testColor = {
		    typename: 'RGBColor',
		    red: '0',
		    green: '0',
		    blue: '0'
	    };

	    var testOpactiy = 100;
	    expect(createjs.getColor(testColor, testOpactiy)).toBe("'rgba(0,0,0,1)'");

	    testColor = {
			    typename: 'RGBColor',
			    red: '255',
			    green: '255',
			    blue: '255'
	    };
	    expect(createjs.getColor(testColor, testOpactiy)).toBe("'rgba(255,255,255,1)'");

	    testColor = {
			    typename: 'CMYKColor',
			    cyan: '0',
			    black: '100',
			    magenta: '0',
			    yellow: '0'
	    }
	    expect(createjs.getColor(testColor, testOpactiy)).toBe("'rgba(0,0,0,1)'");
    });
	
	it('testGetStrokeStyle', function() {
		var pathItem = {
	        "typename" : "PathItem",
	        "closed" : true,
	        "filled" : false,
	        "fillColor" : {
	            "gray" : 0,
	            "typename" : "GrayColor"
	        },
	        "opacity" : 100,
	        "stroked" : true,
	        "strokeColor" : {
	            "cyan" : 0,
	            "magenta" : 0,
	            "yellow" : 0,
	            "black" : 100,
	            "typename" : "CMYKColor"
	        },
	        "strokeDashes" : [

	        ],
	        "strokeDashOffset" : 0,
	        "strokeCap" : "StrokeCap.BUTTENDCAP",
	        "strokeJoin" : "StrokeJoin.MITERENDJOIN",
	        "strokeMiterLimit" : 10,
	        "strokeWidth" : 6,
	        "pathPoints" : [{
	                "anchor" : [
	                    0,
	                    0
	                ],
	                "leftDirection" : [
	                    0,
	                    0
	                ],
	                "rightDirection" : [
	                    0,
	                    0
	                ]
	            }, {
	                "anchor" : [
	                    10,
	                    0
	                ],
	                "leftDirection" : [
	                    10,
	                    0
	                ],
	                "rightDirection" : [
	                    10,
	                    0
	                ]
	            }, {
	                "anchor" : [
	                    10,
	                    10
	                ],
	                "leftDirection" : [
	                    10,
	                    10
	                ],
	                "rightDirection" : [
	                    10,
	                    10
	                ]
	            }, {
	                "anchor" : [
	                    0,
						10
	                ],
	                "leftDirection" : [
	                     0,
						10
	                ],
	                "rightDirection" : [
	                     0,
						10
	                ]
	            }
	        ]
	    };
		
		expect(createjs.getStrokeStyle(pathItem)).toBe("6,0,0,10");
	});


    it('testConvertBitToBase64', function () {
	    var bits = ["000000", "000001", "111111000000"];
	    var base64 = ["A", "B", "/A"];
	    for(var i = 0; i < bits.length; i++) {
		    expect(createjs.convertBitToBase64(bits[i])).toBe(base64[i]);
	    }
    });

    it('testConvertNumberToBits', function () {
	    var number = [0,0.2,0.4,0.6,0.8,10.0,100.0,200000.0,-0.1,-2.0];
	    var bits12 = ["000000000000","000000000010","000000000100","000000000110","000000001000",
	                  "000001100100","001111101000","overflow","100000000001","100000010100"];
	    var bits18 = ["000000000000000000","000000000000000010","000000000000000100",
	                  "000000000000000110","000000000000001000","000000000001100100",
	                  "000000001111101000","overflow","100000000000000001","100000000000010100"];
	    for(var i = 0; i < number.length; i++) {
		    expect(createjs.convertNumberToBits(number[i],12)).toBe(bits12[i]);
		    expect(createjs.convertNumberToBits(number[i],18)).toBe(bits18[i]);
	    }
    });
});


spec/pixitiny.spec.js
// hack for executing code on node.js
if(typeof module !== 'undefined') {
    fs = require('fs');
	var code = fs.readFileSync('js/parser/createjs.js', 'utf-8');
	eval(code);
    code = fs.readFileSync('js/parser/pixitiny.js', 'utf-8');
    eval(code);
}


// actual test-case
describe("PIXI_tiny Parser Test Suite", function() {

	var pixitiny = new Pixitiny();
	
	it("testParsePathItem", function () {
		pixitiny.cropBox = [0,0];
	    var pathItem = {
	        "typename" : "PathItem",
	        "closed" : true,
	        "filled" : false,
	        "fillColor" : {
	            "gray" : 0,
	            "typename" : "GrayColor"
	        },
	        "opacity" : 100,
	        "stroked" : true,
	        "strokeColor" : {
	            "cyan" : 0,
	            "magenta" : 0,
	            "yellow" : 0,
	            "black" : 100,
	            "typename" : "CMYKColor"
	        },
	        "strokeDashes" : [

	        ],
	        "strokeDashOffset" : 0,
	        "strokeCap" : {},
	        "strokeJoin" : {},
	        "strokeMiterLimit" : 10,
	        "strokeWidth" : 6,
	        "pathPoints" : [{
	                "anchor" : [
	                    0,
	                    0
	                ],
	                "leftDirection" : [
	                    0,
	                    0
	                ],
	                "rightDirection" : [
	                    0,
	                    0
	                ]
	            }, {
	                "anchor" : [
	                    10,
	                    0
	                ],
	                "leftDirection" : [
	                    10,
	                    0
	                ],
	                "rightDirection" : [
	                    10,
	                    0
	                ]
	            }, {
	                "anchor" : [
	                    10,
	                    10
	                ],
	                "leftDirection" : [
	                    10,
	                    10
	                ],
	                "rightDirection" : [
	                    10,
	                    10
	                ]
	            }, {
	                "anchor" : [
	                    0,
						10
	                ],
	                "leftDirection" : [
	                     0,
						10
	                ],
	                "rightDirection" : [
	                     0,
						10
	                ]
	            }
	        ]
	    }
		
		expect(pixitiny.parsePathItem(pathItem)).toBe(".ss(6).s(0x000000).p('" + 
			pixitiny.convertPathItemToPathInstruction(pathItem) +"').es()");
	});
	
	it("testGetColor", function () {
	    var testColor = {
		    typename: 'RGBColor',
		    red: '0',
		    green: '0',
		    blue: '0'
	    };

	    var testOpactiy = 100;
	    expect(pixitiny.getColor(testColor, testOpactiy)).toBe("0x000000");

	    testColor = {
			    typename: 'RGBColor',
			    red: '255',
			    green: '255',
			    blue: '255'
	    };
	    expect(pixitiny.getColor(testColor, testOpactiy)).toBe("0xffffff");

	    testColor = {
			    typename: 'CMYKColor',
			    cyan: '0',
			    black: '100',
			    magenta: '0',
			    yellow: '0'
	    }
	    expect(pixitiny.getColor(testColor, testOpactiy)).toBe("0x000000");
    });
	
	it("testConvertToHex", function() {
		var number = [0,0.2,0.4,0.6,0.8,10.0,100.0,200000.0,-0.1,-2.0];
	    var bits12 = ["000000000000","000000000010","000000000100","000000000110","000000001000",
	                  "000001100100","001111101000","overflow","100000000001","100000010100"];
	    var bits18 = ["000000000000000000","000000000000000010","000000000000000100",
	                  "000000000000000110","000000000000001000","000000000001100100",
	                  "000000001111101000","overflow","100000000000000001","100000000000010100"];
	    for(var i = 0; i < number.length; i++) {
		    expect(pixitiny.convertNumberToBits(number[i],12)).toBe(bits12[i]);
		    expect(pixitiny.convertNumberToBits(number[i],18)).toBe(bits18[i]);
		}
	});

});

js/ext.js
function onClickButtonSelect() {
	// @TODO Text aus dem TextFeld in die Zwischenablage kopieren.
	var textField = document.getElementById("output");
	textField.select();
}

function onClickWrap() {
	var textField = document.getElementById("output");
	var checkboxWrap = document.getElementById("wrap");
	if(checkboxWrap.checked) {
		textField.wrap = hard;
	} else {
		textField.wrap = soft;
	}
}

function onClickButtonGenerate() {
	// Dropdown HTML Element in einer Variablen hinterlegen
	var scriptTypesElement = document.getElementById("scriptType");
	// Den aktuellen Wert der Variablen auslesen.
	var scriptTypesValue = scriptTypesElement.options[scriptTypesElement.selectedIndex].value;
	// Das Textfeld für die Ausgabe in einer Variablen hinterlegen
	var textField = document.getElementById("output");
	var prefixField = document.getElementById("prefix");
	if(scriptTypesValue == scriptTypesElement.options[0].value) {
		new CSInterface().evalScript("generateCreateJS({hexColor : false })", function(codeString) {
			textField.value = prefixField.value + codeString;
		});
	} else if(scriptTypesValue == scriptTypesElement.options[1].value) {
		new CSInterface().evalScript("generateCreateJS({hexColor : true })", function(codeString) {
			textField.value = prefixField.value + codeString;
		});
	} else if(scriptTypesValue == scriptTypesElement.options[2].value) {
		new CSInterface().evalScript("generatePixiJson()", function(codeString) {
			textField.value = codeString;
		})
	}
}

function onLoaded() {
	themeManager.init();
}

js/eventHandler.js
function onClickButtonSelect() {
	var textField = document.getElementById("output");
	textField.select();
}

function onClickWrap() {
	var textField = document.getElementById("output");
	var checkboxWrap = document.getElementById("wrap");
	if(checkboxWrap.checked) {
		textField.wrap = hard;
	} else {
		textField.wrap = soft;
	}
}

function onClickButtonGenerate() {
	var scriptTypesElement = document.getElementById("scriptType");
	var scriptTypesValue = scriptTypesElement.options[scriptTypesElement.selectedIndex].value;
	var textField = document.getElementById("output");
	var prefixField = document.getElementById("prefix");
	new CSInterface().evalScript("generateJSON()", function(json) {
		if(scriptTypesValue == scriptTypesElement.options[0].value) {
			textField.value = prefixField.value + (new Createjs()).generate(JSON.parse(json));
		} else if(scriptTypesValue == scriptTypesElement.options[1].value) {
			textField.value = prefixField.value + (new Pixitiny()).generate(JSON.parse(json));
		} else if(scriptTypesValue == scriptTypesElement.options[2].value) {
			textField.value = (new Pixigraphics()).generate(JSON.parse(json));
		} else if(scriptTypesValue == scriptTypesElement.options[3].value) {
			textField.value = json;
		}
	});
}

function onLoaded() {
	themeManager.init();
}

js/themeManager.js
var themeManager = (function () {
  'use strict'; 
 
  /* Convert the Color object to string in hexadecimal format; */
  function toHex(color, delta) {
    function computeValue(value, delta) {
      var computedValue = !isNaN(delta) ? value + delta : value;
      if (computedValue < 0) {
        computedValue = 0;
      } else if (computedValue > 255) {
        computedValue = 255;
      }            
      computedValue = Math.floor(computedValue);
      computedValue = computedValue.toString(16);
      return computedValue.length === 1 ? "0" + computedValue : computedValue;
    } 
    var hex = "";
    if (color) {
      hex = computeValue(color.red, delta) + computeValue(color.green, delta) + computeValue(color.blue, delta);
    }
    return hex;
  }
 
  function addRule(stylesheetId, selector, rule) {
    var stylesheet = document.getElementById(stylesheetId);   
    if (stylesheet) {
      stylesheet = stylesheet.sheet;
      if (stylesheet.addRule) {
        stylesheet.addRule(selector, rule);
      } else if (stylesheet.insertRule) {
        stylesheet.insertRule(selector + ' { ' + rule + ' }', stylesheet.cssRules.length);
      }
    }
  }
 
  /* Update the theme with the AppSkinInfo retrieved from the host product. */
  function updateThemeWithAppSkinInfo(appSkinInfo) {
    // console.log(appSkinInfo)
    var panelBgColor = appSkinInfo.panelBackgroundColor.color;
    var bgdColor = toHex(panelBgColor);
    var fontColor = "F0F0F0";
    if (panelBgColor.red > 122) {
      fontColor = "000000";
    }
 
    var styleId = "hostStyle";
    addRule(styleId, "body", "background-color:" + "#" + bgdColor);
    addRule(styleId, "body", "color:" + "#" + fontColor);
 
    var isLight = appSkinInfo.panelBackgroundColor.color.red >= 127;
    if (isLight) {
      $("#theme").attr("href", "css/light.css");
    } else {
      $("#theme").attr("href", "css/dark.css");
    }
  }
 
  function onAppThemeColorChanged(event) {
    var skinInfo = JSON.parse(window.__adobe_cep__.getHostEnvironment()).appSkinInfo;
    updateThemeWithAppSkinInfo(skinInfo);
  }
 
  function init() {   
    var csInterface = new CSInterface();
    updateThemeWithAppSkinInfo(csInterface.hostEnvironment.appSkinInfo);
    csInterface.addEventListener(CSInterface.THEME_COLOR_CHANGED_EVENT, onAppThemeColorChanged);
  }
 
  return {
    init: init
  };
 
}());

js/parser/pixitiny.js
var Pixitiny = function () {};

Pixitiny.prototype = Object.create(Createjs.prototype);

/**
* Generates a PIXI_tiny API String from a givin PathItem Object.
*
* @param pathItem
* @return PIXI_tiny API String
*/
Pixitiny.prototype.parsePathItem = function (pathItem) {
    var instruction = "";
    var pathPoints = pathItem.pathPoints;

    if (pathItem.filled) {
        instruction += ".f(" + this.getColor(pathItem.fillColor, pathItem.opacity) + ")";
    }
    if (pathItem.stroked) {
		instruction += ".ss(" +pathItem.strokeWidth + ")";
        instruction += ".s(" + this.getColor(pathItem.strokeColor, pathItem.opacity) + ")";
    }
    if (pathPoints.length > 0) {
        instruction += ".p('" + this.convertPathItemToPathInstruction(pathItem) + "')";
    }
    if (pathItem.stroked) {
        //End stroke
        instruction += ".es()";
    }
    if (pathItem.filled) {
        //End fill
        instruction += ".ef()";
    }
    return instruction;
}

/**
* Converts an color object to a PIXI_tiny compatible hex value.
*
* @param color an color object in the format {typename: "RGBColor", red: 0-255, green: 0-255, blue: 0-255} or
* {typename: "CMYKColor", cyan: 0-1, yellow: 0-1, magenta: 0-1, black: 0-1}
* @param opacity the alpha component of the rgba string. 0 is transparent, 1 is opaque.
* @return hex value in the format "0xFFFFFF"
*/
Pixitiny.prototype.getColor = function (color, opacity) {
    var r;
	var g;
	var b;
    var a = opacity / 100;

    if (color.typename == "RGBColor") {
        r = color.red;
        g = color.green;
        b = color.blue;
    } else if (color.typename == "CMYKColor") {
        //CYMK Color Codes
        var k = color.black;
        var c = color.cyan;
        var m = color.magenta;
        var y = color.yellow;

        //RGB Color Codes
        r = 255 * (1 - c / 100) * (1 - k / 100);
        g = 255 * (1 - m / 100) * (1 - k / 100);
        b = 255 * (1 - y / 100) * (1 - k / 100);
    }
    return "0x" + this.convertToHex(r) + this.convertToHex(g) + this.convertToHex(b);
}

/*
 * Convers a number to a two letter hex value
 */
Pixitiny.prototype.convertToHex = function (number) {
    var hex = Math.round(number).toString(16);
    if (hex.length < 2) {
        hex = "0" + hex;
    }
    return hex;
}

js/parser/createjs.js
var Createjs = function () {};

Createjs.prototype.BASE_64 = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N",
    "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "a", "b", "c", "d", "e", "f", "g",
    "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z",
    "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "+", "/"];

//Map instructions to values
Createjs.prototype.instructions = {
    "moveTo" : "000",
    "lineTo" : "001",
    "quadraticCurveTo" : "010",
    "bezierCurveTo" : "011",
    "closePath" : "100"
};

//Map instructions to parameter count
Createjs.prototype.paramCount = {
    "moveTo" : 2,
    "lineTo" : 2,
    "quadraticCurveTo" : 4,
    "bezierCurveTo" : 6,
    "closePath" : 0
};

//Map size of position property to bits
Createjs.prototype.sizeOfPositions = {
    "12" : 0,
    "18" : 1
};

Createjs.prototype.unusedBits = "00";

Createjs.prototype.cropBox;


/**
* Generates a CreateJS Tiny API String from a giving JSON-Object.
*
* @param json
* @return CreateJS Tiny API String
*/
Createjs.prototype.generate = function (json) {
    var objects = json.selection;
    var instruction = "";
    this.cropBox = json.cropBox;
    for (var i = objects.length - 1; i >= 0; i--) {
        if (objects[i].typename == "PathItem") {
            instruction += this.parsePathItem(objects[i]);
        } else if (objects[i].typename == "GroupItem") {
            instruction += this.parseGroupItem(objects[i]);
        } else if (objects[i].typename == "CompoundPathItem") {
            instruction += this.parseCompoundPathItem(objects[i]);
        }
    }
    if (objects.length > 0) {
        instruction += ".cp()";
    }
    return instruction;
}

/**
* Generates a CreateJS Tiny API String from a givin PathItem Object.
*
* @param pathItem
* @return CreateJS Tiny API String
*/
Createjs.prototype.parsePathItem = function (pathItem) {
    var instruction = "";
    var pathPoints = pathItem.pathPoints;

    if (pathItem.filled) {
        instruction += ".f(" + this.getColor(pathItem.fillColor, pathItem.opacity) + ")";
    }
    if (pathItem.stroked) {
        instruction += ".ss(" + this.getStrokeStyle(pathItem) + ")";
        instruction += ".s(" + this.getColor(pathItem.strokeColor, pathItem.opacity) + ")";
    }
    if (pathItem.stroked) {
        var strokeDashes = pathItem.strokeDashes;
        var strokeDashOffset = 0;
        instruction += ".sd([" + strokeDashes + "]," + strokeDashOffset + ")";
    }
    if (pathPoints.length > 0) {
        instruction += ".p('" + this.convertPathItemToPathInstruction(pathItem) + "')";
    }
    if (pathItem.stroked) {
        //End stroke
        instruction += ".es()";
    }
    if (pathItem.filled) {
        //End fill
        instruction += ".ef()";
    }
    return instruction;
}

Createjs.prototype.parseGroupItem = function (groupItem) {
    var instruction = "";
    for (var i = 0; i < groupItem.groupItems.length; i++) {
        instruction += this.parseGroupItem(groupItem.groupItems[i]);
    }
    for (var i = 0; i < groupItem.pathItems.length; i++) {
        instruction += this.parsePathItem(groupItem.pathItems[i]);
    }
    for (var i = 0; i < groupItem.compoundPathItems.length; i++) {
        instruction += this.parseCompoundPathItem(groupItem.compoundPathItems[i]);
    }
    return instruction;
}

Createjs.prototype.parseCompoundPathItem = function (compoundPathItem) {
    var instruction = "";
    for (var i = 0; i < compoundPathItem.pathItems.length; i++) {
        instruction += this.parsePathItem(compoundPathItem.pathItems[i]);
    }
    return instruction;
}

/**
* Generates a encoded path string by converting a series of pathPoints to an bit string, which is
* then encoded to a BASE_64 string.
*
* @param pathItem
* @return a encoded path string
*/
Createjs.prototype.convertPathItemToPathInstruction = function (pathItem) {

    var pathPoints = pathItem.pathPoints;

    //Easeljs uses absolute movement for the first command
    var moveToX = pathPoints[0].anchor[0] + this.cropBox[0];
    var moveToY = -1 * pathPoints[0].anchor[1] + this.cropBox[1];

    //12 bits you can used numbers up to 2047, 18 bits you can use numbers up to 131071.
    var sizeOfPosition = 18;

    var bitInstruction = this.instructions["moveTo"] + this.sizeOfPositions[sizeOfPosition] +
        this.unusedBits + this.convertNumberToBits(moveToX, sizeOfPosition) +
        this.convertNumberToBits(moveToY, sizeOfPosition);

    var pathInstruction = this.convertBitToBase64(bitInstruction);

    if (pathPoints.length > 1) {
        for (var i = 0; i < (pathItem.closed ? pathPoints.length : (pathPoints.length - 1)); i++) {
            var previousPoint = pathPoints[i % pathPoints.length];
            var currentPoint = pathPoints[(i + 1) % pathPoints.length];

            if (currentPoint.anchor[0] == currentPoint.leftDirection[0] &&
					currentPoint.anchor[0] == currentPoint.rightDirection[0] &&
					currentPoint.anchor[1] == currentPoint.leftDirection[1] &&
					currentPoint.anchor[1] == currentPoint.rightDirection[1]) {
                var lineToX = currentPoint.anchor[0] - previousPoint.anchor[0];
                var lineToY = -1 * currentPoint.anchor[1] + previousPoint.anchor[1];

                bitInstruction = this.instructions["lineTo"] + this.sizeOfPositions[sizeOfPosition] + this.unusedBits +
                    this.convertNumberToBits(lineToX, sizeOfPosition) + this.convertNumberToBits(lineToY, sizeOfPosition);

                pathInstruction += this.convertBitToBase64(bitInstruction);
            } else {
                var anchorX = currentPoint.anchor[0] - currentPoint.leftDirection[0];
                var anchorY = -1 * currentPoint.anchor[1] + currentPoint.leftDirection[1];

                var leftDirectionX = currentPoint.leftDirection[0] - previousPoint.rightDirection[0];
                var leftDirectionY = -1 * currentPoint.leftDirection[1] + previousPoint.rightDirection[1];

                var rightDirectionX = previousPoint.rightDirection[0] - previousPoint.anchor[0];
                var rightDirectionY = -1 * previousPoint.rightDirection[1] + previousPoint.anchor[1];

                bitInstruction = this.instructions["bezierCurveTo"] +
                    this.sizeOfPositions[sizeOfPosition] + this.unusedBits +
                    this.convertNumberToBits(rightDirectionX, sizeOfPosition) +
                    this.convertNumberToBits(rightDirectionY, sizeOfPosition) +
                    this.convertNumberToBits(leftDirectionX, sizeOfPosition) +
                    this.convertNumberToBits(leftDirectionY, sizeOfPosition) +
                    this.convertNumberToBits(anchorX, sizeOfPosition) +
                    this.convertNumberToBits(anchorY, sizeOfPosition);

                pathInstruction += this.convertBitToBase64(bitInstruction);
            }
        }
    }
    return pathInstruction;
}

/**
* Converts an color object to a css compatible rgba string.
*
* @param color an color object in the format {typename: "RGBColor", red: 0-255, green: 0-255, blue: 0-255} or
* {typename: "CMYKColor", cyan: 0-1, yellow: 0-1, magenta: 0-1, black: 0-1}
* @param opacity the alpha component of the rgba string. 0 is transparent, 1 is opaque.
* @return a css compatible color string in the format "rgba(255,255,255,1)"
*/
Createjs.prototype.getColor = function (color, opacity) {
    var r,
    g,
    b;
    var a = opacity / 100;

    if (color.typename == "RGBColor") {
        r = color.red;
        g = color.green;
        b = color.blue;
    } else if (color.typename == "CMYKColor") {
        //CYMK Color Codes
        var k = color.black;
        var c = color.cyan;
        var m = color.magenta;
        var y = color.yellow;

        //RGB Color Codes
        r = 255 * (1 - c / 100) * (1 - k / 100);
        g = 255 * (1 - m / 100) * (1 - k / 100);
        b = 255 * (1 - y / 100) * (1 - k / 100);
    }
    return "'rgba(" + Math.round(r) + "," + Math.round(g) + "," + Math.round(b) + "," + a + ")'";
}

/**
* Filters needed properties from a pathItem and converts them to arguments for the ss method.
*
* @return ss arguments
*/
Createjs.prototype.getStrokeStyle = function (pathItem) {
    var strokeWidth = pathItem.strokeWidth;
    var strokeCap;
    var strokeJoin;
    var strokeMiterLimit = pathItem.strokeMiterLimit;

    if (pathItem.strokeCap == "StrokeCap.BUTTENDCAP") {
        strokeCap = 0;
    } else if (pathItem.strokeCap == "StrokeCap.PROJECTINGENDCAP") {
        strokeCap = 1;
    } else if (pathItem.strokeCap == "StrokeCap.ROUNDENDCAP") {
        strokeCap = 2;
    }

    if (pathItem.strokeJoin == "StrokeJoin.MITERENDJOIN") {
        strokeJoin = 0;
    } else if (pathItem.strokeJoin == "StrokeJoin.ROUNDENDJOIN") {
        strokeJoin = 1;
    } else if (pathItem.strokeJoin == "StrokeJoin.BEVELENDJOIN") {
        strokeJoin = 2;
    }
    return strokeWidth + "," + strokeCap + "," + strokeJoin + "," + strokeMiterLimit;
}

/*
 * Converts a string of bits to a base64 string
 */
Createjs.prototype.convertBitToBase64 = function (bits) {
    var base64String = "";
    var loops = bits.length / 6;
    for (var i = 0; i < loops; i++) {
        var bitsChunk = bits.substring(i * 6, i * 6 + 6);
        var intFromBitsChunk = parseInt(bitsChunk, 2);
        var base64Char = this.BASE_64[intFromBitsChunk];
        base64String += base64Char;
    }
    return base64String;
}

/**
 * Floors a number an then converts it to a bit String. When the number is negative the MSB will
 * be 1, when it is positive the MSB will be 0.
 *
 * @param number Number which will be converted to a bit String
 * @param size Number of bits which will be used to represent number
 * @return bit string
 */
Createjs.prototype.convertNumberToBits = function (number, size) {
    //Easeljs has an accuracy of 1/10 of a pixel.
    number = number * 10;
    var sign = (number < 0 ? "1" : "0");
    var result = Math.abs(Math.round(number)).toString(2);
    while (result.length < size - 1) {
        result = "0" + result;
    }
	if(result.length - 1 > size) {
		return "overflow";
	}
    return sign + result;
}